### PR TITLE
Add validation logic for `Dims.last_used` to prevent slider selection errors when rolling dims

### LINKED ISSUE #2388: Slider focus issues ?
## 🐛 Bug

I'm not too sure how to reproduce as I was smashing keys to test the action manager with `python examples/nD_volume.py`
but it seem like some combinaison of roll/transpose/3d<->2d and Alt-Up/Down can trigger an error:

```
Traceback (most recent call last):
  File "/Users/mbussonnier/dev/vispy/vispy/app/backends/_qt.py", line 520, in keyPressEvent
    self._keyEvent(self._vispy_canvas.events.key_press, ev)
  File "/Users/mbussonnier/dev/vispy/vispy/app/backends/_qt.py", line 569, in _keyEvent
    func(native=ev, key=key, text=str(ev.text()), modifiers=mod)
  File "/Users/mbussonnier/dev/vispy/vispy/util/event.py", line 454, in __call__
    self._invoke_callback(cb, event)
  File "/Users/mbussonnier/dev/vispy/vispy/util/event.py", line 472, in _invoke_callback
    _handle_exception(self.ignore_callback_errors,
  File "/Users/mbussonnier/dev/vispy/vispy/util/event.py", line 470, in _invoke_callback
    cb(event)
  File "/Users/mbussonnier/dev/napari/napari/utils/key_bindings.py", line 629, in on_key_press
    self.press_key(combo)
  File "/Users/mbussonnier/dev/napari/napari/utils/key_bindings.py", line 585, in press_key
    gen = func()
  File "/Users/mbussonnier/dev/napari/napari/components/_viewer_key_bindings.py", line 66, in focus_axes_up
  File "/Users/mbussonnier/dev/napari/napari/components/dims.py", line 293, in _focus_up
    index = (sliders.index(self.last_used) + 1) % len(sliders)
ValueError: X is not in list
```
Where X is 0, 3 .. and likely 1 or 2 as well, though I haven't observed those.

### PR DESCRIPTION
**Pull request by dalthviz**
# Fixes/Closes

Closes #2388 

# Description

Missing validations for the `last_used` value caused no slider selection when rolling dims and errors when trying to change the focused slider through shortcuts.

A preview with the changes here:

roll_dims

# References

Discussion of possible solutions over the issue comments

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# How has this been tested?

- [x] new test added
- [x] all tests pass with my change
- [x] I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.

### CODE REVIEW

File: napari/_qt/widgets/_tests/test_qt_dims.py
Comment (andy-sweet) [inline napari/_qt/widgets/_tests/test_qt_dims.py]: Optional: I think there's a preference for not showing widgets in tests, so you might consider using `QWidget.isVisibleTo(view)` instead of `QWidget.isVisible`, in which case you shouldn't need `view.show()`.

File: napari/components/_tests/test_viewer_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_tests/test_viewer_mouse_bindings.py:41]: Why did these values change?

I think it's because the test sets up with `last_used == 2`, which was allowed before, but is now invalid and this PR now instead causes `last_used` to be `0`. So now the simulated ctrl+scroll events cause us to traverse points in dimension 0 and not 2. Is that correct? If so, I think you should explicitly set `last_used = 0` in the test to make it more clear.

It also makes me think about the slightly magic behavior here. Overall I think it's better than erroring and I don't think we should warn all the time because this could happen a lot with normal usage. But maybe we should warn when using the API directly (rather than through the GUI)?

Don't do anything for now, but would be good to get a second opinion on this PR. @jni often has a good intuition for these things.

File: napari/components/dims.py
Comment (andy-sweet) [inline napari/components/dims.py]: Not certain, but I think you might want to use `range_` or equivalently `updated['range']` here instead, so we handle validation of multiple values on initialization.

File: napari/components/dims.py
Comment (andy-sweet) [inline napari/components/dims.py]: Optional: mild preference for pulling out `nsteps = cls._nsteps_from_range(dims_range)` because it avoids a little extra compute and is slightly easier to read non-nested/chained expressions.

File: napari/components/_tests/test_dims.py
Comment (andy-sweet) [inline napari/components/_tests/test_dims.py:345]: As the `last_used` setter does some validation now, it would be good to assert that setting the valid value works as expected.

```suggestion
    dims.last_used = 1
    assert dims.last_used == 1
```

That also sets an informative precondition before taking the action of changing `ndisplay`.

**Review by andy-sweet — CHANGES_REQUESTED** [review]
Generally looks good to me, though I have a few minor requests/comments/questions. After that I'm happy to approve.

**Comment by andy-sweet** [comment]
Even though this PR is small and contained, it would be good to get a second opinion here, as I was involved in the suggesting this type of solution in #2388.

Maybe one of @jni or @psobolewskiPhD would be good additional reviewers?

File: napari/components/_tests/test_viewer_mouse_bindings.py
Comment (dalthviz) [reply napari/components/_tests/test_viewer_mouse_bindings.py:41]: You are correct! Checked locally and indeed the `last_used` value fallbacks to 0 after setting it to 2 in the test. Also, the addition of a warning makes sense to me :+1: it's true that is not so trivial to understand why, after `last_used` is explicitly set, it can inmediatly have another value (could even make users think there is an issue if not informed about the validation and fallback to a valid value)

Comment (andy-sweet) [reply napari/components/_tests/test_viewer_mouse_bindings.py:41]: In that case, I think we should at least update the test with `viewer.dims.last_used = 0`.

I don't feel strongly about the warning, especially because the `Dims` root validator does a lot of this type of coercion without warning.

**Review by andy-sweet — APPROVED** [review]
Looks good to me now!

**Comment by andy-sweet** [comment]
@psobolewskiPhD : added you as a second reviewer as you were able to reproduce #2388 and were part of the discussion there. This PR is pretty small and self-contained, but let me know if you don't have time to review right now.

File: napari/components/_tests/test_viewer_mouse_bindings.py
Comment (psobolewskiPhD) [reply napari/components/_tests/test_viewer_mouse_bindings.py:41]: I don't have much intuition regarding this, sorry.
On the GUI side, the behavior seems logical to me (other than clicking on the roll dims button stealing keyboard focus from sliders, but that's a different issue/PR).
The bugfix code makes sense to me, so I think this is OK and we can revisit if needed?

**Review by psobolewskiPhD — APPROVED** [review]
Sorry for the delay...
The GUI behavior seems intuitive to me and the bugfix code makes sense, so I think 👍

**Pull request merged**
